In [1]:
import pandas as pd

## 정리
- df.read_csv
- df.info()
> - df.columns
> - df.index
- df
> - value 확인
> - head(), tail()

<br>

다음 시간: 숫자 항목과, 문자 항목에서 각각 확인해야할 사항들

- read_csv() : csv 파일 불러오는 함수
- csv : comma seperated value

In [2]:
# 고객별 계약 1건이 1 sample.
df1 = pd.read_csv("01_Contract_Data.csv")
df1

Index  Member_ID Sales_Type Contract_Type Channel    Datetime  Term  \
0          1   66758234         렌탈          일반계약    영업방판  2019-05-06    60   
1          2   66755948         렌탈          교체계약    영업방판  2020-02-20    60   
2          3   66756657         렌탈          일반계약  홈쇼핑/방송  2019-02-28    60   
3          4   66423450        멤버십        멤버십3유형     재계약  2019-05-13    12   
4          5   66423204        멤버십        멤버십3유형     재계약  2019-05-10    12   
...      ...        ...        ...           ...     ...         ...   ...   
51296  51298   66579515         렌탈        프로모션계약   대형마트A  2019-03-01    60   
51297  51299   66799558         렌탈          일반계약   대형마트A  2019-04-01    60   
51298  51300   66799197         렌탈        프로모션계약    영업방판  2019-04-01    39   
51299  51301   66792778         렌탈          일반계약  홈쇼핑/방송  2020-02-06    60   
51300  51302   66799607         렌탈          일반계약  홈쇼핑/방송  2019-04-24    60   

      Payment_Type Product_Type  Amount_Month Customer_Type   Age Address1  \
0              CMS        DES-1         96900            개인  42.0      경기도   
1             카드이체        DES-1        102900            개인  39.0      경기도   
2              CMS        DES-1         96900            개인  48.0      경기도   
3              CMS        DES-1         66900            개인  39.0      경기도   
4              CMS        DES-1         66900            개인  60.0      경기도   
...            ...          ...           ...           ...   ...      ...   
51296          CMS       DES-3A         96900            개인  47.0      경기도   
51297          CMS        DES-1         96900            개인  42.0      경기도   
51298         카드이체          ERA        120900            개인  65.0    서울특별시   
51299         카드이체        DES-1         96900            개인  54.0    서울특별시   
51300          CMS        DES-1         96900            개인  53.0    서울특별시   

      Address2 State  Overdue_count Overdue_Type Gender  Credit_Rank    Bank  
0          경기도  계약확정              0           없음     여자          9.0   새마을금고  
1          경기도  계약확정              0           없음     남자          2.0    현대카드  
2          경기도  계약확정              0           없음     여자          8.0    우리은행  
3          경기도  계약확정              0           없음     남자          5.0  농협회원조합  
4          경기도  기간만료             12           있음     남자          8.0  농협회원조합  
...        ...   ...            ...          ...    ...          ...     ...  
51296      경기도  계약확정              0           없음     남자          NaN    기업은행  
51297      경기도  계약확정              0           없음     여자          8.0   새마을금고  
51298    서울특별시  계약확정              0           없음     여자          1.0    롯데카드  
51299    서울특별시  계약확정              0           없음     여자          2.0    롯데카드  
51300    서울특별시  계약확정              0           없음     여자          8.0    신한은행  

[51301 rows x 20 columns]

- structured data이므로 index, column, value를 확인한다.
- 51301 rows × 20 columns에서 index와 column을 확인한다. 
- 함수로 이를 확인할 수 있다.

In [3]:
# df.columns() : df의 column 확인
df1.columns

Index(['Index', 'Member_ID', 'Sales_Type', 'Contract_Type', 'Channel',
       'Datetime', 'Term', 'Payment_Type', 'Product_Type', 'Amount_Month',
       'Customer_Type', 'Age', 'Address1', 'Address2', 'State',
       'Overdue_count', 'Overdue_Type', 'Gender', 'Credit_Rank', 'Bank'],
      dtype='object')

In [4]:
# df.index() : df의 index 확인
df1.index

RangeIndex(start=0, stop=51301, step=1)

In [5]:
# df.info() : df의 index와 column(+ 항목별 dtype)을 한번에 확인하는 함수 +
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51301 entries, 0 to 51300
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Index          51301 non-null  int64  
 1   Member_ID      51301 non-null  int64  
 2   Sales_Type     51301 non-null  object 
 3   Contract_Type  51301 non-null  object 
 4   Channel        51301 non-null  object 
 5   Datetime       51301 non-null  object 
 6   Term           51301 non-null  int64  
 7   Payment_Type   51301 non-null  object 
 8   Product_Type   51301 non-null  object 
 9   Amount_Month   51301 non-null  int64  
 10  Customer_Type  51299 non-null  object 
 11  Age            44329 non-null  float64
 12  Address1       51299 non-null  object 
 13  Address2       51299 non-null  object 
 14  State          51301 non-null  object 
 15  Overdue_count  51301 non-null  int64  
 16  Overdue_Type   51301 non-null  object 
 17  Gender         51301 non-null  object 
 18  Credit

분석가들은 데이터의 구조와 타입을 확인할 때, df.info()를 사용한다.
- 타입별 항목 수
- 데이터의 전체적인 구조를 info()로 확인

간단하게 value를 확인하기 위해 head() 또는 tail()을 사용한다.

## number, object 항목에서 확인해야하는 사항들
- 엑셀처럼 다 보여주는 게 아니라 생략된다. (...)
- 각 항목에 어떤 값이 있는지 정확하게 확인하는 방법


<br>

- 숫자 데이터와 문자 데이터는 dtype이 다르다. 
- dtype이 다르면 통계적 분석 방법이 다르다.
- 따라서 숫자 데이터와 문자 데이터에서 기본적으로 확인해야 할 사항이 서로 다르다.


- 처음부터 다시 실행해줘야한다. 이 때, kernel/run all을 클릭한다.
- 앞 코드를 사용하기 위해서는 다시 실행해야 한다.

### 숫자데이터 확인하기
실무에서 확인하는 사항들
- 평균: 숫자데이터의 대표값(중앙값)을 확인 (오만건의 데이터를 대표할 수 있는 숫자 1개)
- 합 : 숫자데이터의 전체 합 확인 (e.g 매출액)

- 숫자 데이터가 굉장히 많은데, 이거를 일일이 다 확인해야 하는가?
> 다 확인하면 좋다. 하지만 수가 너무 많으면 일일이 확인하기 어렵다.
> 이 때, describe()를 사용해서 확인한다.
> 숫자 데이터 다 갖고와서 요약 통계량(평균, 표준편차, 최소값, 최대값, ... )을 확인해준다. 

- 합을 확인할 때는 df1.sum(axis=1)

In [8]:
# 내 코드
import numpy as np
df1.select_dtypes(include=np.number)

Index  Member_ID  Term  Amount_Month   Age  Overdue_count  Credit_Rank
0          1   66758234    60         96900  42.0              0          9.0
1          2   66755948    60        102900  39.0              0          2.0
2          3   66756657    60         96900  48.0              0          8.0
3          4   66423450    12         66900  39.0              0          5.0
4          5   66423204    12         66900  60.0             12          8.0
...      ...        ...   ...           ...   ...            ...          ...
51296  51298   66579515    60         96900  47.0              0          NaN
51297  51299   66799558    60         96900  42.0              0          8.0
51298  51300   66799197    39        120900  65.0              0          1.0
51299  51301   66792778    60         96900  54.0              0          2.0
51300  51302   66799607    60         96900  53.0              0          8.0

[51301 rows x 7 columns]

In [9]:
df1["Term"] # df1의 Term항목 Ser로 가져오기

0        60
1        60
2        60
3        12
4        12
         ..
51296    60
51297    60
51298    39
51299    60
51300    60
Name: Term, Length: 51301, dtype: int64

In [11]:
# Term : 계약기간
df1["Term"].mean() # 평균값 확인 : 오만명 고객들이 평균적으로 55개월 제품을 렌탈을 하는구나

55.63914933432097

In [12]:
# 고객 1명당 월별 납부 비용
df1["Amount_Month"]

0         96900
1        102900
2         96900
3         66900
4         66900
          ...  
51296     96900
51297     96900
51298    120900
51299     96900
51300     96900
Name: Amount_Month, Length: 51301, dtype: int64

In [14]:
df1["Amount_Month"].sum() # 데이터의 합 확인 : 전체 계약금은 48억 2천만원이 월 계약 수익이다.

4822036176

In [15]:
df1.describe()

Index     Member_ID          Term   Amount_Month           Age  \
count  51301.000000  5.130100e+04  51301.000000   51301.000000  44329.000000   
mean   25651.703612  6.266432e+07     55.639149   93994.974289     50.024093   
std    14809.828628  1.216146e+07     12.009915   15304.263988     10.983877   
min        1.000000  2.568798e+07     12.000000   54603.000000     25.000000   
25%    12826.000000  6.643148e+07     60.000000   81900.000000     42.000000   
50%    25652.000000  6.676578e+07     60.000000   96900.000000     49.000000   
75%    38477.000000  6.678160e+07     60.000000   98400.000000     57.000000   
max    51302.000000  6.696986e+07     60.000000  215700.000000    102.000000   

       Overdue_count   Credit_Rank  
count   51301.000000  42520.000000  
mean        0.161381      3.428810  
std         1.122193      2.213453  
min         0.000000      0.000000  
25%         0.000000      1.000000  
50%         0.000000      3.000000  
75%         0.000000      5.000000  
max        15.000000     10.000000

In [18]:
# 문자형 sum은 append된게 합이다.
df1.sum()

C:\Users\USER\AppData\Local\Temp\ipykernel_13576\1283687288.py:2: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.sum()


Index                                                   1315958047
Member_ID                                            3214742058621
Sales_Type       렌탈렌탈렌탈멤버십멤버십렌탈렌탈렌탈렌탈렌탈렌탈렌탈렌탈렌탈렌탈멤버십렌탈렌탈멤버십렌탈멤버...
Contract_Type    일반계약교체계약일반계약멤버십3유형멤버십3유형프로모션계약패키지계약패키지계약일반계약프로...
Channel          영업방판영업방판홈쇼핑/방송재계약재계약영업방판홈쇼핑/방송홈쇼핑/방송영업방판영업방판홈쇼...
Datetime         2019-05-062020-02-202019-02-282019-05-132019-0...
Term                                                       2854344
Payment_Type     CMS카드이체CMSCMSCMSCMS카드이체카드이체카드이체CMS카드이체카드이체카드이체...
Product_Type     DES-1DES-1DES-1DES-1DES-1DES-3ADES-1DES-2DES-3...
Amount_Month                                            4822036176
Age                                                      2217518.0
State            계약확정계약확정계약확정계약확정기간만료계약확정계약확정계약확정계약확정계약확정계약확정계약...
Overdue_count                                                 8279
Overdue_Type     없음없음없음없음있음없음없음없음없음없음없음없음없음없음없음없음없음없음없음없음없음없음없음...
Gender           여자남자여자남자남자여자남자남자여자여자여자여자여자남자남자여자남자여자여자여자여자남자여

In [19]:
# 최고령은 몇이고, 등등을 확인한다.

In [22]:
df1.select_dtypes(include=np.number).sum(axis=0)

Index            1.315958e+09
Member_ID        3.214742e+12
Term             2.854344e+06
Amount_Month     4.822036e+09
Age              2.217518e+06
Overdue_count    8.279000e+03
Credit_Rank      1.457930e+05
dtype: float64

## 문자 데이터 확인하기
- 항목 : 어떤 항목이 있는가?
- 빈도수 : 그 항목이 몇개씩 구성되어 있는가?

- 한 항목에 몇가지 unique 값이 있는지 detail하게 확인하려면 df1.value_counts()를 사용해야 한다
- 기본적인, 전반적인 정보만 확인하려면 df1.describe(include="object)를 사용한다.

In [23]:
df1["Sales_Type"] # ser형태로 문자 데이터를 뽑는다 : object

0         렌탈
1         렌탈
2         렌탈
3        멤버십
4        멤버십
        ... 
51296     렌탈
51297     렌탈
51298     렌탈
51299     렌탈
51300     렌탈
Name: Sales_Type, Length: 51301, dtype: object

In [24]:
# Sales_Type 항목에 어떤 고유 value들이 있는지 보여준다
df1["Sales_Type"].unique() 

array(['렌탈', '멤버십'], dtype=object)

In [25]:
# 어떤 은행을 사용했는가?
df1["Bank"].unique()

array(['새마을금고', '현대카드', '우리은행', '농협회원조합', '농협중앙회', '롯데카드', '신한은행', '국민카드',
       '부산은행', '우체국', '국민은행', '기업은행', 'SC제일은행', nan, '대구은행', '하나은행',
       '신한카드', '삼성카드', '광주은행', '외환은행', 'BC카드', '씨티은행', '신협중앙회', '경남은행',
       '미래에셋증권', '외환카드', '수협중앙회', '유안타증권', '전북은행', '상호저축은행', '대신증권',
       '산업은행', '현대증권', '산림조합중앙회', '제주은행', '삼성증권', '하나대투증권', 'SK증권',
       '수협카드', '한국투자증권', '신한금융투자', '하나SK', 'NH농협카드', '메리츠종합금융증권',
       '하이투자증권', '우리카드', '대우증권', '씨티카드'], dtype=object)

In [27]:
# 렌탈은 몇건, 멤버십은 몇건 있는가? : 빈도수
df1["Sales_Type"].value_counts() # 문자 데이터 항목의 빈도수 계산

렌탈     46483
멤버십     4818
Name: Sales_Type, dtype: int64

In [31]:
# 각 은행별로 몇명이 사용했는가? : 빈도수
df1["Bank"].value_counts()

롯데카드         9516
농협회원조합       6277
국민은행         6002
농협중앙회        3899
신한은행         3522
우리은행         3386
기업은행         1963
신한카드         1532
하나은행         1446
국민카드         1311
BC카드         1264
새마을금고         964
부산은행          889
삼성카드          884
현대카드          876
대구은행          746
우체국           718
외환은행          586
외환카드          530
경남은행          442
SC제일은행        439
광주은행          347
신협중앙회         341
전북은행          195
씨티은행          162
수협중앙회         160
제주은행           40
유안타증권          27
산업은행           23
현대증권           11
삼성증권            7
하나SK            6
미래에셋증권          5
한국투자증권          4
신한금융투자          4
NH농협카드          4
우리카드            3
대우증권            2
산림조합중앙회         1
대신증권            1
하나대투증권          1
SK증권            1
수협카드            1
상호저축은행          1
메리츠종합금융증권       1
하이투자증권          1
씨티카드            1
Name: Bank, dtype: int64

In [33]:
# 숫자의 describe 같은 함수가 문자 데이터에는 없는가? 
# 문자 데이터만 가져와서 기본적인 요약 통계량 확인 : 문자 데이터의 요약 통계량 확인

# count: 개수
# unique: unique value의 개수
# top : unique value 중 가장(top) 빈도 수가 높은 object type의 value -> 렌탈이라는 값이 제일 많다. (최고 빈도수를 갖는 값)
# freq : 가장 빈도수가 많은 unique value의 frequency만 출력

df1.describe(include="object")

Sales_Type Contract_Type Channel    Datetime Payment_Type Product_Type  \
count       51301         51301   51301       51301        51301        51301   
unique          2             9      16         577            5            6   
top            렌탈        프로모션계약    영업방판  2019-01-31          CMS        DES-1   
freq        46483         15811   23767        1167        32825        39133   

       Customer_Type Address1 Address2  State Overdue_Type Gender   Bank  
count          51299    51299    51299  51301        51301  51301  48542  
unique             2        8       14      4            2      2     47  
top               개인      경기도      경기도   계약확정           없음     여자   롯데카드  
freq           46263    18353    14883  50620        49110  35602   9516